### PR TITLE
Initial support for SQLCMD variables in sql project tree

### PR DESCRIPTION
**Pull request by kisantia**
Note: not for the January release, so this will get checked in after code complete

This adds support for displaying the SQLCMD variables in a sql project in the project tree. Add, Edit, and Delete functionality will be added in a separate PR after the swap is made to use the project parsing in DacFx, instead of the current TS code.

image

image

### CODE REVIEW

**Review by SakshiS-harma — COMMENTED** [review]
Is TestProject file added by mistake?

**Comment by kisantia** [comment]
> Is TestProject file added by mistake?

Removed it - thanks for catching it! Looks like it was generated by a test, so I'll send out a PR to fix that

File: extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	private sqlcmdVariables: SqlCmdVariableTreeItem[] = [];
+
+	constructor(project: ProjectRootTreeItem) {
+		super(vscode.Uri.file(path.join(project.projectUri.fsPath, constants.sqlcmdVariablesNodeName)), project);
+
+		this.construct();
+	}
+
+	private construct() {
+		for (const sqlCmdVariable of Object.keys((this.parent as ProjectRootTreeItem).project.sqlCmdVariables)) {
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts:27]: please add handling for possible undefined scenarios here.

**Review by alanrenmsft — APPROVED** [review]

File: extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		this.construct();
+	}
+
+	private construct() {
+		if (!(this.parent as ProjectRootTreeItem).project.sqlCmdVariables) {
+			return;
+		}
+
+		for (const sqlCmdVariable of Object.keys((this.parent as ProjectRootTreeItem).project.sqlCmdVariables)) {
+			if (sqlCmdVariable) {
```
Comment (Benjin) [inline extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts:32]: Can `Object.keys()` return any null/undefined objects?

File: extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	private sqlcmdVariables: SqlCmdVariableTreeItem[] = [];
+
+	constructor(project: ProjectRootTreeItem) {
+		super(vscode.Uri.file(path.join(project.projectUri.fsPath, constants.sqlcmdVariablesNodeName)), project);
+
+		this.construct();
+	}
+
+	private construct() {
+		if (!(this.parent as ProjectRootTreeItem).project.sqlCmdVariables) {
```
Comment (Benjin) [inline extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts:27]: Might be worth extract this to a variable for readability and reuse below

**Review by Benjin — APPROVED** [review]

**Review by SakshiS-harma — APPROVED** [review]

File: extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts (hunk 2 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/models/tree/sqlcmdVariableTreeItem.ts:34]: I don't think so, just an empty array if there aren't any keys

**Pull request merged**